api/src/utils/keycloak-utils.ts

```diff
@@ -71,20 +140,63 @@ export const coerceUserIdentitySource = (userIdentitySource: string | null): SYS
 };
 
 /**
- * Parses out the user's identifier from a keycloak token.
+ * Returns the user information identifier (aka: username).
  *
- * @example getUserIdentifier({ ....token, bceid_username: 'jsmith@idir' }) => 'jsmith'
+ * @example getUserIdentifier({ idir_username: 'jsmith' }) => 'jsmith'
+ *
+ * @param {KeycloakUserInformation} keycloakUserInformation
+ * @return {*}  {string}
+ */
+export const getUserIdentifier = (keycloakUserInformation: KeycloakUserInformation): string => {
+  if (isIdirUserInformation(keycloakUserInformation)) {
+    return keycloakUserInformation.idir_username;
+  }
+
+  if (isBceidBusinessUserInformation(keycloakUserInformation) || isBceidBasicUserInformation(keycloakUserInformation)) {
+    return keycloakUserInformation.bceid_username;
+  }
+
+  return keycloakUserInformation.username;
+};
+
+/**
+ * Get a `KeycloakUserInformation` object from a Keycloak Bearer Token (IDIR or BCeID Basic or BCeID Business token).
  *
  * @param {Record<string, any>} keycloakToken
- * @return {*} {(string | null)}
+ * @return {*}  {(KeycloakUserInformation | null)}
  */
-export const getUserIdentifier = (keycloakToken: Record<string, any>): string | null => {
-  const userIdentifier =
-    keycloakToken?.['idir_username'] || keycloakToken?.['bceid_username'] || keycloakToken?.['sims_system_username'];
+export const getKeycloakUserInformationFromKeycloakToken = (
+  keycloakToken: Record<string, any>
+): KeycloakUserInformation | null => {
+  const result = KeycloakUserInformation.safeParse(keycloakToken);
 
-  if (!userIdentifier) {
+  if (!result.success) {
     return null;
   }
 
-  return userIdentifier;
+  return result.data;
+};
+
+export const isIdirUserInformation = (
+  keycloakUserInformation: KeycloakUserInformation
+): keycloakUserInformation is IdirUserInformation => {
+  return keycloakUserInformation.identity_provider === SYSTEM_IDENTITY_SOURCE.IDIR.toLowerCase();
+};
+
+export const isBceidBasicUserInformation = (
+  keycloakUserInformation: KeycloakUserInformation
+): keycloakUserInformation is BceidBasicUserInformation => {
+  return keycloakUserInformation.identity_provider === SYSTEM_IDENTITY_SOURCE.BCEID_BASIC.toLowerCase();
+};
+
+export const isBceidBusinessUserInformation = (
+  keycloakUserInformation: KeycloakUserInformation
+): keycloakUserInformation is BceidBusinessUserInformation => {
+  return keycloakUserInformation.identity_provider === SYSTEM_IDENTITY_SOURCE.BCEID_BUSINESS.toLowerCase();
+};
+
+export const isDatabaseUserInformation = (
+  keycloakUserInformation: KeycloakUserInformation
+): keycloakUserInformation is DatabaseUserInformation => {
+  return keycloakUserInformation.identity_provider === SYSTEM_IDENTITY_SOURCE.DATABASE.toLowerCase();
```

Comment:
I'm assuming keycloak currently provides this value as a lowercase string. But is there a possible edge case where we may expect `keycloakUserInformation.identity_provider` to be lowercase but it isn't? That would cause this check to erronesouly fail

Reply:
I included a call to lowercase in the zod schema, so if used correctly, this identity_provider field should always be lowercase already